File: src/main/java/net/zomis/chess/ChessBoard.java
package net.zomis.chess;

import java.util.stream.IntStream;

import static net.zomis.chess.Piece.PAWN;

public class ChessBoard {

    private static final int SIZE = 8;
    private final ChessTile[][] tiles;
    private Player current = Player.WHITE;
    private MoveEngine moveEngine = new MoveEngine(this);

    public ChessBoard() {
        this.tiles = new ChessTile[SIZE][SIZE];
        for (int y = 0; y < SIZE; y++) {
            for (int x = 0; x < SIZE; x++) {
                tiles[y][x] = new ChessTile(x, y);
            }
        }
    }

    public void move(String move) {
        System.out.println("Move: " + move);
        ChessPosition destination = ChessPosition.fromMoveNotation(move);
        ChessTile source = null;
        if (move.length() == 2) {
            // Moving a pawn
            source = IntStream.range(0, SIZE)
                .mapToObj(i -> new ChessPosition(destination.getX(), i))
                .map(this::tile)
                .filter(ch -> ch.getOwner() == current)
                .filter(ch -> ch.getPiece() == PAWN)
                .collect(new UniqueCollector<>())
                .orElseThrow(() -> new IllegalStateException("Unexpected number of pawns found for move " + move));
        }
        if (moveEngine.moveValid(source, destination)) {
            this.move(source, destination);
        } else {
            throw new IllegalStateException("Illegal move: " + move);
        }

        switchTurn();
    }

    private void switchTurn() {
        current = current.opponent();
    }

    public ChessBoard moves(String moves) {
        ChessBoard board = ChessBoard.standardGame();
        for (String move : moves.split(" ")) {
            board.move(move);
        }
        return board;
    }

    public ChessTile tile(char column, int row) {
        if (row <= 0 || row > SIZE || column < 'A' || column >= 'A' + SIZE) {
            throw new IndexOutOfBoundsException("Column " + column + " and row " +
                row + " is out of bounds for size" + SIZE);
        }
        return tiles[row - 1][column - 'A'];
    }

    public static ChessBoard standardGame() {
        ChessBoard board = new ChessBoard();
        for (int i = 0; i < SIZE; i++) {
            board.tile(col(i), 2).set(Player.WHITE, PAWN);
            board.tile(col(i), SIZE - 1).set(Player.BLACK, PAWN);
        }

        for (Player pl : Player.values()) {
            int row = pl == Player.WHITE ? 1 : SIZE;
            board.tile('A', row).set(pl, Piece.ROOK);
            board.tile('B', row).set(pl, Piece.KNIGHT);
            board.tile('C', row).set(pl, Piece.BISHOP);
            board.tile('D', row).set(pl, Piece.QUEEN);
            board.tile('E', row).set(pl, Piece.KING);
            board.tile('F', row).set(pl, Piece.BISHOP);
            board.tile('G', row).set(pl, Piece.KNIGHT);
            board.tile('H', row).set(pl, Piece.ROOK);
        }

        return board;
    }

    private static char col(int i) {
        return (char) ('A' + i);
    }

    public void print(Player perspective) {
        for (int y = 1; y <= SIZE; y++) {
            for (int x = 1; x <= SIZE; x++) {
                ChessTile tile = tileFromPerspective(x, y, perspective);
                String str = tile.toString();
                System.out.print(str);
                System.out.print(' ');
            }
            System.out.println();
        }
    }

    private ChessTile tileFromPerspective(int x, int y, Player perspective) {
        int cx = x - 1;
        int cy = y - 1;
        if (perspective == Player.WHITE) {
            cy = SIZE - y;
        }
        if (perspective == Player.BLACK) {
            cx = SIZE - x;
        }
        if (!(new ChessPosition(cx, cy).onBoard())) {
            throw new IllegalArgumentException("not on board: " + x + ", " + y + " perspective " + perspective
            + ". values " + cx + ", " + cy);
        }
        return tiles[cy][cx];
    }

    public ChessTile tile(ChessPosition pos) {
        return this.tiles[pos.getY()][pos.getX()];
    }

    void move(ChessTile tile, ChessPosition pos) {
        System.out.println("Move " + tile.getFullName() + " at " + tile.getPosition() + " to " + pos);
        tile(pos).set(tile);
        tile.clear();
    }
}

File: chess-core/src/main/kotlin/net/zomis/chess/core/ChessBoard.kt
package net.zomis.chess.core

class ChessBoard {

    val SIZE = 8

    private val tiles: Array<Array<ChessTile>>
    private var current = Player.WHITE
    private val moveEngine = MoveEngine(this)

    init {
        this.tiles = Array(SIZE) { y ->
            Array(SIZE) { x ->
                ChessTile(x, y)
            }
        }
    }

    fun move(move: String) {
        println("Move: " + move)
        val destination = ChessPosition.fromMoveNotation(move)
        var source: ChessTile? = null
        if (move.length == 2) {
            // Moving a pawn
            val list = (0 until SIZE)
                    .map({ i -> ChessPosition(destination.x, i) })
                    .map({ this.tile(it) })
                    .filter({ ch -> ch.owner == current })
                    .filter({ ch -> ch.piece == Piece.PAWN })
                    .toList()
            if (list.size == 1) {
                source = list[0]
            } else {
                throw IllegalStateException("Unexpected number of pawns found for move " + move)
            }
        }
        if (moveEngine.moveValid(source, destination)) {
            this.move(source, destination)
        } else {
            throw IllegalStateException("Illegal move: " + move)
        }

        switchTurn()
    }

    private fun switchTurn() {
        current = current.opponent()
    }

    fun moves(moves: String): ChessBoard {
        val board = ChessBoard.standardGame()
        for (move in moves.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
            board.move(move)
        }
        return board
    }

    fun tile(column: Char, row: Int): ChessTile {
        if (row <= 0 || row > SIZE || column < 'A' || column >= 'A' + SIZE) {
            throw IndexOutOfBoundsException("Column " + column + " and row " +
                    row + " is out of bounds for size" + SIZE)
        }
        return tiles[row - 1][column - 'A']
    }

    fun print(perspective: Player) {
        for (y in 1..SIZE) {
            for (x in 1..SIZE) {
                val tile = tileFromPerspective(x, y, perspective)
                val str = tile.toString()
                print(str)
                print(' ')
            }
            println()
        }
    }

    private fun tileFromPerspective(x: Int, y: Int, perspective: Player): ChessTile {
        var cx = x - 1
        var cy = y - 1
        if (perspective == Player.WHITE) {
            cy = SIZE - y
        }
        if (perspective == Player.BLACK) {
            cx = SIZE - x
        }
        if (!ChessPosition(cx, cy).onBoard()) {
            throw IllegalArgumentException("not on board: " + x + ", " + y + " perspective " + perspective
                    + ". values " + cx + ", " + cy)
        }
        return tiles[cy][cx]
    }

    fun tile(pos: ChessPosition): ChessTile {
        return this.tiles[pos.y][pos.x]
    }

    internal fun move(tile: ChessTile?, pos: ChessPosition) {
        println("Move " + tile!!.fullName + " at " + tile.position + " to " + pos)
        tile(pos).set(tile)
        tile.clear()
    }

    companion object {

        private val SIZE = 8

        fun standardGame(): ChessBoard {
            val board = ChessBoard()
            for (i in 0 until SIZE) {
                board.tile(col(i), 2).set(Player.WHITE, Piece.PAWN)
                board.tile(col(i), SIZE - 1).set(Player.BLACK, Piece.PAWN)
            }

            for (pl in Player.values()) {
                val row = if (pl == Player.WHITE) 1 else SIZE
                board.tile('A', row).set(pl, Piece.ROOK)
                board.tile('B', row).set(pl, Piece.KNIGHT)
                board.tile('C', row).set(pl, Piece.BISHOP)
                board.tile('D', row).set(pl, Piece.QUEEN)
                board.tile('E', row).set(pl, Piece.KING)
                board.tile('F', row).set(pl, Piece.BISHOP)
                board.tile('G', row).set(pl, Piece.KNIGHT)
                board.tile('H', row).set(pl, Piece.ROOK)
            }

            return board
        }

        private fun col(i: Int): Char {
            return ('A' + i).toChar()
        }
    }
}
